File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/NonTableManagerSamples/OpenPositionNetting/src/OrderMonitorNetting.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;
using fxcore2;

namespace OrdersMonitor
{
    /// <summary>
    /// Helper class for monitoring creation open positions using a open order.
    /// On no dealing desk more than one position can be create. It is depends on
    /// liquidity on forex market, The class stores all open positions 
    /// </summary>
    internal class OrderMonitorNetting
    {
        static public bool IsOpeningOrder(O2GOrderRow order)
        {
            return order.Type.StartsWith("O");
        }

        static public bool IsClosingOrder(O2GOrderRow order)
        {
            return order.Type.StartsWith("C");
        }


        private enum OrderState
        {
            OrderExecuting,
            OrderExecuted,
            OrderCanceled,
            OrderRejected
        }
        private volatile OrderState mState;

        private List<O2GTradeRow> mTrades;
        private List<O2GTradeRow> mUpdatedTrades;
        private List<O2GClosedTradeRow> mClosedTrades;
        private volatile int mTotalAmount;
        private volatile int mRejectAmount;
        private O2GOrderRow mOrder;
        private string mRejectMessage;
        private int mInitialAmount;

        public enum ExecutionResult
        {
            Executing,
            Executed,
            PartialRejected,
            FullyRejected,
            Canceled
        };


        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="order">Order for monitoring of execution</param>
        public OrderMonitorNetting(O2GOrderRow order) : this(order, 0)
        {
        }

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="order">Order for monitoring of execution</param>
        /// <param name="iNetPositionAmount">Initial amount of trade</param>
        public OrderMonitorNetting(O2GOrderRow order, int iNetPositionAmount)
        {
            mOrder = order;
            mTrades = new List<O2GTradeRow>();
            mUpdatedTrades = new List<O2GTradeRow>();
            mClosedTrades = new List<O2GClosedTradeRow>();
            mState = OrderState.OrderExecuting;
            mResult = ExecutionResult.Executing;
            mInitialAmount = iNetPositionAmount;
            mTotalAmount = 0;
            mRejectAmount = 0;
            mRejectMessage = "";
        }

        /// <summary>
        /// Process trade adding during order execution
        /// </summary>
        public void OnTradeAdded(O2GTradeRow tradeRow)
        {
            string sTradeOrderID = tradeRow.OpenOrderID;
            string sOrderID = mOrder.OrderID;
            if (sTradeOrderID.Equals(sOrderID))
            {
                mTrades.Add(tradeRow);

                if (mState == OrderState.OrderExecuted ||
                    mState == OrderState.OrderRejected ||
                    mState == OrderState.OrderCanceled)
                {
                    if (IsAllTradesReceived())
                        SetResult(true);
                }
            }
        }

        /// <summary>
        /// Process order data changing during execution
        /// </summary>
        public void OnOrderChanged(O2GOrderRow orderRow)
        {
            //STUB
        }

        /// <summary>
        /// Process order deletion as result of execution
        /// </summary>
        public void OnOrderDeleted(O2GOrderRow orderRow)
        {
            string sDeletedOrderID = orderRow.OrderID;
            string sOrderID = mOrder.OrderID;
            if (sDeletedOrderID.Equals(sOrderID))
            {
                // Store Reject amount
                if (OrderRowStatus.Rejected.Equals(orderRow.Status))
                {
                    mState = OrderState.OrderRejected;
                    mRejectAmount = orderRow.Amount;
                    mTotalAmount = orderRow.OriginAmount - mRejectAmount;
                    if (!string.IsNullOrEmpty(mRejectMessage) && IsAllTradesReceived())
                        SetResult(true);
                }
                else if (OrderRowStatus.Canceled.Equals(orderRow.Status))
                {
                    mState = OrderState.OrderCanceled;
                    mRejectAmount = orderRow.Amount;
                    mTotalAmount = orderRow.OriginAmount - mRejectAmount;
                    if (IsAllTradesReceived())
                        SetResult(false);
                }
                else
                {
                    mRejectAmount = 0;
                    mTotalAmount = orderRow.OriginAmount;
                    mState = OrderState.OrderExecuted;
                    if (IsAllTradesReceived())
                        SetResult(true);
                }
            }
        }

        /// <summary>
        /// Process reject message as result of order execution
        /// </summary>
        public void OnMessageAdded(O2GMessageRow messageRow)
        {
            if (mState == OrderState.OrderRejected ||
                mState == OrderState.OrderExecuting)
            {
                bool IsRejectMessage = CheckAndStoreMessage(messageRow);
                if (mState == OrderState.OrderRejected && IsRejectMessage)
                    SetResult(true);
            }
        }

        /// <summary>
        /// Process trade updating during order execution
        /// </summary>
        public void OnTradeUpdated(O2GTradeRow tradeRow)
        {
            string sTradeOrderID = tradeRow.OpenOrderID;
            string sOrderID = mOrder.OrderID;
            if (sTradeOrderID.Equals(sOrderID))
            {
                mUpdatedTrades.Add(tradeRow);
                if (mState == OrderState.OrderExecuted ||
                        mState == OrderState.OrderRejected ||
                        mState == OrderState.OrderCanceled)
                {
                    if (IsAllTradesReceived())
                    {
                        SetResult(true);
                    }
                }
            }
        }

        /// <summary>
        /// Process trade closing during order execution
        /// </summary>
        public void OnClosedTradeAdded(O2GClosedTradeRow closedTradeRow)
        {
            string sOrderID = mOrder.OrderID;
            string sClosedTradeOrderID = closedTradeRow.CloseOrderID;
            if (sClosedTradeOrderID.Equals(sOrderID))
            {
                mClosedTrades.Add(closedTradeRow);

                if (mState == OrderState.OrderExecuted ||
                    mState == OrderState.OrderRejected ||
                    mState == OrderState.OrderCanceled)
                {
                    if (IsAllTradesReceived())
                        SetResult(true);
                }
            }
        }

        /// <summary>
        /// Event about order execution is completed and all affected trades as opened/closed, all reject/cancel processed
        /// </summary>
        public event EventHandler OrderCompleted;

        /// <summary>
        /// Result of Order execution
        /// </summary>
        public ExecutionResult Result
        {
            get
            {
                return mResult;
            }
        }
        private volatile ExecutionResult mResult;

        /// <summary>
        /// Order execution is completed (with any result)
        /// </summary>
        public bool IsOrderCompleted
        {
            get
            {
                return (mResult != ExecutionResult.Executing);
            }
        }

        /// <summary>
        /// Monitored order
        /// </summary>
        public O2GOrderRow Order
        {
            get
            {
                return mOrder;
            }
        }

        /// <summary>
        /// List of Trades which were opened as effects of order execution
        /// </summary>
        public ReadOnlyCollection<O2GTradeRow> Trades
        {
            get
            {
                return mTrades.AsReadOnly();
            }
        }

        /// <summary>
        /// List of Trades which were updated as effects of order execution
        /// </summary>
        public ReadOnlyCollection<O2GTradeRow> UpdatedTrades
        {
            get
            {
                return mUpdatedTrades.AsReadOnly();
            }
        }

        /// <summary>
        /// List of Trades which were closed as effects of order execution
        /// </summary>
        public ReadOnlyCollection<O2GClosedTradeRow> ClosedTrades
        {
            get
            {
                return mClosedTrades.AsReadOnly();
            }
        }

        /// <summary>
        /// Amount of rejected part of order
        /// </summary>
        public int RejectAmount
        {
            get
            {
                return mRejectAmount;
            }
        }

        /// <summary>
        /// Info message with a reason of reject
        /// </summary>
        public string RejectMessage
        {
            get
            {
                return mRejectMessage;
            }
        }


        private void SetResult(bool bSuccess)
        {
            if (bSuccess)
            {
                if (mRejectAmount == 0)
                    mResult = ExecutionResult.Executed;
                else
                    mResult = (mTrades.Count == 0 && mClosedTrades.Count == 0) ?  ExecutionResult.FullyRejected : ExecutionResult.PartialRejected;
            }
            else
                mResult = ExecutionResult.Canceled;

            if (OrderCompleted != null)
                OrderCompleted(this, EventArgs.Empty);

        }

        private bool IsAllTradesReceived()
        {
            if (mState == OrderState.OrderExecuting)
                return false;
            int iCurrentTotalAmount = 0;
            for (int i = 0; i < mTrades.Count; i++)
            {
                iCurrentTotalAmount += mTrades[i].Amount;
            }
            for (int i = 0; i < mUpdatedTrades.Count; i++)
            {
                iCurrentTotalAmount += mUpdatedTrades[i].Amount;
            }
            for (int i = 0; i < mClosedTrades.Count; i++)
            {
                iCurrentTotalAmount += mClosedTrades[i].Amount;
            }
            return Math.Abs(iCurrentTotalAmount - mInitialAmount) == mTotalAmount;
        }

        private bool CheckAndStoreMessage(O2GMessageRow message)
        {
            string sFeature = message.Feature;
            if (MessageFeature.MarketCondition.Equals(sFeature))
            {
                string sText = message.Text;
                int findPos = sText.IndexOf(mOrder.OrderID);
                if (findPos > -1)
                {
                    mRejectMessage = sText;
                    return true;
                }
            }
            return false;
        }

    }

    internal class OrderRowStatus
    {
        public static string Rejected = "R";
        public static string Canceled = "C";
        public static string Executed = "F";
        //...
    }

    internal class MessageFeature
    {
        public static String MarketCondition = "5";
        //...
    }

}

File: samples/Python/common_samples/OrderMonitorNetting.py
# Copyright 2019 Gehtsoft USA LLC

# Licensed under the license derived from the Apache License, Version 2.0 (the "License"); 
# you may not use this file except in compliance with the License.

# You may obtain a copy of the License at

# http://fxcodebase.com/licenses/open-source/license.html

# Unless required by applicable law or agreed to in writing, software 
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import List
from enum import Enum

from forexconnect import fxcorepy


class OrderMonitorNetting:

    class ExecutionResult(Enum):
        EXECUTING = 1
        EXECUTED = 2
        PARTIAL_REJECTED = 3
        FULLY_REJECTED = 4
        CANCELED = 5

    class OrderState(Enum):
        ORDER_EXECUTING = 1
        ORDER_EXECUTED = 2
        ORDER_CANCELED = 3
        ORDER_REJECTED = 4
    
    __market_condition = "5"

    def __init__(self, order: fxcorepy.O2GOrderRow, i_net_position_amount: int = 0) -> None:
        self.__order = order
        self.__trades = []
        self.__updated_trades = []
        self.__closed_trades = []
        self.__state = OrderMonitorNetting.OrderState.ORDER_EXECUTING
        self.__result = OrderMonitorNetting.ExecutionResult.EXECUTING
        self.__total_amount = 0
        self.__reject_amount = 0
        self.__reject_message = ""
        self.__initial_amount = i_net_position_amount

    @staticmethod
    def is_opening_order(order: fxcorepy.O2GOrderRow) -> bool:
        return order.type.startswith("O")

    @staticmethod
    def is_closing_order(order: fxcorepy.O2GOrderRow) -> bool:
        return order.type.startswith("C")

    # Process trade adding during order execution
    def on_trade_added(self, trade: fxcorepy.O2GTradeRow) -> None:
        trade_order_id = trade.open_order_id
        order_id = self.__order.order_id
        if trade_order_id == order_id:
            self.__trades.append(trade)
            if self.__state == OrderMonitorNetting.OrderState.ORDER_EXECUTED or \
                    self.__state == OrderMonitorNetting.OrderState.ORDER_REJECTED or \
                    self.__state == OrderMonitorNetting.OrderState.ORDER_CANCELED:
                if self.is_all_trades_received:
                    self.set_result(True)

    # Process trade updating during order execution
    def on_trade_updated(self, trade_row: fxcorepy.O2GTradeRow) -> None:
        s_trade_order_id = trade_row.open_order_id
        s_order_id = self.__order.order_id
        if s_trade_order_id == s_order_id:
            self.__updated_trades.append(trade_row)
            if self.__state == OrderMonitorNetting.OrderState.ORDER_EXECUTED or \
                    self.__state == OrderMonitorNetting.OrderState.ORDER_REJECTED or \
                    self.__state == OrderMonitorNetting.OrderState.ORDER_CANCELED:
                if self.is_all_trades_received:
                    self.set_result(True)

    # Process trade closing during order execution
    def on_closed_trade_added(self, closed_trade: fxcorepy.O2GClosedTradeRow) -> None:
        order_id = self.__order.order_id
        closed_trade_order_id = closed_trade.close_order_id
        if order_id == closed_trade_order_id:
            self.__closed_trades.append(closed_trade)
            if self.__state == OrderMonitorNetting.OrderState.ORDER_EXECUTED or \
                    self.__state == OrderMonitorNetting.OrderState.ORDER_REJECTED or \
                    self.__state == OrderMonitorNetting.OrderState.ORDER_CANCELED:
                if self.is_all_trades_received:
                    self.set_result(True)

    # Process order deletion as result of execution
    def on_order_deleted(self, order: fxcorepy.O2GOrderRow) -> None:
        deleted_order_id = order.order_id
        order_id = self.__order.order_id
        if deleted_order_id == order_id:
            # Store Reject amount
            if order.Status.startswith("R"):
                self.__state = OrderMonitorNetting.OrderState.ORDER_REJECTED
                self.__reject_amount = order.amount
                self.__total_amount = order.origin_amount - self.__reject_amount
                if self.__reject_message != "" and self.is_all_trades_received:
                    self.set_result(True)
            else:
                if order.Status.startswith("C"):
                    self.__state = OrderMonitorNetting.OrderState.ORDER_CANCELED
                    self.__reject_amount = order.amount
                    self.__total_amount = order.origin_amount - self.__reject_amount
                    if self.is_all_trades_received:
                        self.set_result(False)
                else:
                    self.__reject_amount = 0
                    self.__total_amount = order.OriginAmount
                    self.__state = OrderMonitorNetting.OrderState.ORDER_EXECUTED
                    if self.is_all_trades_received:
                        self.set_result(True)

    def on_message_added(self, message: fxcorepy.O2GMessageRow) -> None:
        if self.__state == OrderMonitorNetting.OrderState.ORDER_REJECTED or \
                self.__state == OrderMonitorNetting.OrderState.ORDER_EXECUTING:
            is_reject_message = self.check_and_store_message(message)
            if self.__state == OrderMonitorNetting.OrderState.ORDER_REJECTED and is_reject_message:
                self.set_result(True)

    @property
    def order_row(self) -> fxcorepy.O2GOrderRow:
        return self.__order

    @property
    def trade_rows(self) -> List[fxcorepy.O2GTradeRow]:
        return self.__trades

    @property
    def updated_trade_rows(self) -> List[fxcorepy.O2GTradeRow]:
        return self.__updated_trades

    @property
    def closed_trade_rows(self) -> List[fxcorepy.O2GClosedTradeRow]:
        return self.__closed_trades

    @property
    def reject_amount(self) -> int:
        return self.__reject_amount

    @property
    def reject_message(self) -> str:
        return self.__reject_message

    @property
    def result(self) -> ExecutionResult:
        return self.__result

    @property
    def is_order_completed(self) -> bool:
        return self.__result != OrderMonitorNetting.ExecutionResult.EXECUTING

    def check_and_store_message(self, message: fxcorepy.O2GMessageRow) -> bool:
        feature = message.feature
        if feature == self.__market_condition:
            text = message.text
            if self.__order.order_id in text:
                self.__reject_message = message.text
                return True
        return False

    @property
    def is_all_trades_received(self) -> bool:
        if self.__state == OrderMonitorNetting.OrderState.ORDER_EXECUTING:
            return False
        i_current_total_amount = 0
        for trade in self.__trades:
            i_current_total_amount += trade.amount

        for trade in self.__updated_trades:
            i_current_total_amount += trade.amount
        
        for trade in self.__closed_trades:
            i_current_total_amount += trade.amount

        return abs(i_current_total_amount - self.__initial_amount) == self.__total_amount

    def set_result(self, success: bool) -> None:
        if success:
            if self.__reject_amount == 0:
                self.__result = OrderMonitorNetting.ExecutionResult.EXECUTED
            else:
                self.__result = OrderMonitorNetting.ExecutionResult.FULLY_REJECTED \
                    if (len(self.__trades) == 0 and len(self.__closed_trades) == 0) \
                    else OrderMonitorNetting.ExecutionResult.PARTIAL_REJECTED
            
        else:
            self.__result = OrderMonitorNetting.ExecutionResult.CANCELED
